src/saving/serializers/EntityReferenceConverter.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Reflection;
+using Newtonsoft.Json;
+
+public class EntityReferenceConverter : JsonConverter
+{
+    public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+    {
+        if (value == null)
+        {
+            writer.WriteNull();
+            return;
+        }
+
+        var objectType = value.GetType();
+
+        var genericTypes = objectType.GenericTypeArguments;
+
+        if (genericTypes.Length != 1)
+            throw new JsonException("Invalid generic types for EntityReference");
+
+        // Even though Microbe is used here, it shouldn't matter at all, as we just want to grab the property name
+        var property = objectType.GetProperty(nameof(EntityReference<Microbe>.Value));
+
+        if (property == null)
+            throw new JsonException("Value property not found in EntityReference");
+
+        var internalValue = property.GetValue(value);
+
+        if (internalValue == null)
+        {
+            writer.WriteNull();
+            return;
+        }
```

Comment:
I could see this getting us in trouble. If the code of a game object expects a EntityReference variable not to be null, but allows for that EntityReference to have a null Value (which this design currently allows, even though I don't know why you would do that), then saving and loading that object might break it. Worse, I imagine that that would be a scenario that wouldn't happen often, making the future save bug harder to track down.

Reply:
I'm having a little bit of trouble following your line of reasoning here.

I wrote the serializer this way because I wanted it to be *transparent* to the save. So the saves just plain have the *contents* of the EntityReference saved, and no details about the entity reference at all. This way save compatibility is kept and I think it's nicer this way as the internal helper class doesn't need to be output in a save.

I'm thinking along the lines that a EntityReference should *never* be null, but the value can turn null whenever, so client code using the class needs to be designed with these constraints in mind.

Reply:
I was worried about the case of an EntityReference with a value of null. It sounds like that was intended here, so is the idea that the serializer always assumes Entities are wrapped in an EntityReference, so it will read a null (which I presume doesn't write an empty string) then wrap around that? That's not what I took away from reading this, but that strategy should be safe...

Reply:
> I was worried about the case of an EntityReference with a value of null.

That's very much the default state. And also the state the EntityReference goes in if the thing it refers to is destroyed. So I don't think there's any other way to design it really.

> serializer always assumes Entities are wrapped in an EntityReference

I mean it doesn't really. It just reads in a null (in which case it default constructs an EntityReference) or it reads in an object of T and constructs a EntityReference<T> out of that. So it always returns a valid EntityReference instance, but it may have a null as its Value.

And when writing it works so that it writes a null if the EntityReference is null or the value is null. Otherwise it just serializes the held Value.

As I said I designed the serialization of EntityReference to be entirely transparent for the save format.